Aller Anfang ist schwer: Ansätze für Green Software-Engineering (#221)
adessoAG/devblog: _posts/2021-11-26-aller-anfang-ist-schwer-ansaetze-fuer-green-software-engineering.md

Bei binären Daten wie Bildern sieht es im Prinzip ähnlich aus.
Diese sollten effizient mit einer angepassten Auflösung und Qualität heruntergeladen werden.
Hier bieten sich Content Delivery Networks an, die diese Aufgabe gut erledigen.

# Skalierbarkeit

Skalierbarkeit wird häufig unter dem Aspekt betrachtet, dass hohe Zugriffszahlen auftreten, die sonst unüblich sind.
In der Regel gibt man sich aber ansonsten zufrieden, wenn die Anwendung schnell reagiert.
Im Sinne von Green Software-Engineering muss Skalierbarkeit aber auch in die andere Richtung betrachtet werden.
Man sollte sich (auch aus Kostengründen) die Frage stellen, ob zwei Instanzen eines Microservices wirklich nötig sind, wenn gerade niemand zugreift?
Reicht dann eine Instanz oder gar keine?

Das Thema Skalierbarkeit wurde in vielerlei Hinsicht durch das Aufkommen von Kubernetes (für die breite Masse) gelöst.
Dabei kann man ganz trivial eine Mindest- und eine Höchstzahl an Pods (Instanzen) pro Service definieren, die dann automatisch skalieren.
Entscheidend für die Technologie ist allerdings die Startzeit eines Microservice und da sind die auf "klassischen" Webframeworks basierenden Anwendungen leider recht langsam.

> Da ist mit einer Startzeit von über zehn Sekunden zu rechnen, in vielen Fällen sogar deutlich mehr. Das ist nicht wirklich kompatibel mit dem Skalierungsgedanken. Natürlich spricht immer noch viel für den Einsatz von Spring Boot u.a., bei diesem Aspekt sind andere Frameworks wie Quarkus, Micronaut und Ktor überlegen (siehe [hier](https://dzone.com/articles/not-only-spring-boot-a-review-of-alternatives)).

Reviewer: ```suggestion
Bei diesen Technologien ist mit einer Startzeit von über zehn Sekunden zu rechnen, in vielen Fällen sogar deutlich mehr. Das ist nicht wirklich kompatibel mit dem Skalierungsgedanken. Natürlich spricht immer noch viel für den Einsatz von Spring Boot u.ä., bei diesem Aspekt sind aber andere Frameworks wie Quarkus, Micronaut und Ktor überlegen (siehe [hier](https://dzone.com/articles/not-only-spring-boot-a-review-of-alternatives)).
```
Spring Boot kann beispielsweise mit GraalVM ähnlich effizient gestaltet werden (https://blog.codecentric.de/en/2020/05/spring-boot-graalvm/) vielleicht kannst du da noch in einem (Neben-)Satz drauf eingehen.

# Caching

> Webanwendungen verbrauchen z.B. dann viel Speicher, wenn sie Daten cachen müssen. Caching ist prinzipiell aber zu empfehlen, da dadurch gleiche Berechnungen, Datenbank- oder API-Anfragen nicht ständig wiederholt werden müssen. Caching, das sich über mehrere Instanzen eines Microservices erstreckt, ist hier der Königsweg und würde noch stärker vermeiden (insbesondere bei Skalierungen), dass Berechnungen mehrfach gemacht werden.

Reviewer: ```suggestion
Webanwendungen verbrauchen z.B. dann viel Speicher, wenn sie Daten cachen müssen. Caching ist prinzipiell aber zu empfehlen, da dadurch gleiche Berechnungen, Datenbank- oder API-Anfragen nicht ständig wiederholt werden müssen. Caching, das sich über mehrere Instanzen eines Microservice erstreckt, ist hier der beste Weg und würde noch stärker vermeiden (insbesondere bei Skalierungen), dass Berechnungen mehrfach durchgeführt werden.
```

> Ansonsten ist anzumerken, dass Microframeworks wie Quarkus und Ktor im Speicherverbrauch deutlich günstiger sind als beispielsweise Platzhirsch Spring Boot.

Reviewer: ```suggestion
Ansonsten ist anzumerken, dass Microframeworks wie Quarkus und Ktor im Speicherverbrauch deutlich günstiger sind als beispielsweise der Platzhirsch Spring Boot.
```
siehe mein Kommentar zu GraalVM

# Reaktiv entwickeln

> In die Java-Community zog vor ein paar Jahren die nicht-blockierende Entwicklung ein. Der Grund dafür war, dass andere Technologiefamilien wie NodeJS bei der Verarbeitung von Requests eine bessere Performance zeigten und das vor allem am nicht-blockierenden Ansatz lag ([Ein aktueller Beitrag
> dazu ist hier zu sehen](https://da-14.com/blog/nodejs-vs-java-backend-language-large-software-development)).

Reviewer: ```suggestion
In die Java-Community zog vor ein paar Jahren die nicht-blockierende Entwicklung ein.
Der Grund hierfür war, dass andere Technologiefamilien wie NodeJS bei der Verarbeitung von Requests eine bessere Performance zeigten, was vor allem am nicht-blockierenden Ansatz lag ([Ein aktueller Beitrag dazu ist hier zu sehen](https://da-14.com/blog/nodejs-vs-java-backend-language-large-software-development)).
```

> Gleichzeitig benötigen asynchrone Frameworks mehr Speicher und die "Organisation" der nicht-blockierenden Prozesse gilt als aufwändig und damit nicht als Ressourcen schonend.

Reviewer: ```suggestion
Gleichzeitig benötigen asynchrone Frameworks mehr Speicher und die "Organisation" der nicht-blockierenden Prozesse gilt als aufwändig und damit nicht als ressourcenschonend.
```

Zudem hängt der Nutzen stark von der Länge einer Blockade eines Threads ab ([5 Aspekte reativer Programmierung](https://javapro.io/5-aspekte-reaktive-programmierung/)).

> Trotzdem sollte man auch im Java- oder JVM-Umfeld reaktiv entwickeln, da dadurch deutlich schnellere Antwortzeiten zu erwarten sind (siehe z.B. [hier](https://dzone.com/articles/spring-boot-20-webflux-reactive-performance-test)) und das dann in der Regel auch Ressourcen spart (weniger Rechenzeit, weniger Speichernutzung, wenige Wartezeit für den Client).

Reviewer: ```suggestion
Trotzdem sollte man auch im Java- oder JVM-Umfeld reaktiv entwickeln, da dadurch deutlich schnellere Antwortzeiten möglich sind (siehe z.B. [hier](https://dzone.com/articles/spring-boot-20-webflux-reactive-performance-test)) und das dann in der Regel auch Ressourcen spart (weniger Rechenzeit, weniger Speichernutzung, wenige Wartezeit für den Client).
```
Kannst du anstelle von "hier" einen passenden Bezeichner für den Artikel verwenden?

Ein großer Nachteil reaktiver Programmierung in Java ist allerdings das Schreiben und die Lesbarkeit des Codes.

> Viel Code wird mit Funktionen beschrieben, die häufig schwer verständlich sind und leicht den Blick auf das Wesentliche, also die Geschäftslogik, versperren. [Der Empfehlung des Autors](https://nexocode.com/blog/posts/reactive-streams-vs-coroutines/) kann ich mir anschließen, Kotlin Coroutines wirken deutlich einfacher als die gängigen Java-Bibliotheken.

Reviewer: ```suggestion
Viel Code wird mit Funktionen beschrieben, die häufig schwer verständlich sind und leicht den Blick auf das Wesentliche, also die Geschäftslogik, versperren. [Der Empfehlung dieses Autors](https://nexocode.com/blog/posts/reactive-streams-vs-coroutines/) kann ich mir anschließen, Kotlin Coroutines wirken deutlich einfacher als die gängigen Java-Bibliotheken.
```

# Zusammenfassung

> Technologieauswahl und Programmierung spielen eine wichtige Rolle, für den ökologischen Fußabdruck einer Webanwendung.

Reviewer: ```suggestion
Technologieauswahl und Programmierung spielen für den ökologischen Fußabdruck einer Webanwendung eine wichtige Rolle.
```

> Hier wurde exemplarisch einige Beispiele gezeigt, wie man Ressourcen sparen kann, in dem man auf Microframeworks, moderne Serialisierungsbibliotheken, skalierbare Architekturen, reaktive Entwicklung und GraphQL setzt.

Reviewer: ```suggestion
Anhand von Beispielen haben wir gesehen, wie wir Ressourcen sparen können, indem wir auf Microframeworks, moderne Serialisierungsbibliotheken, skalierbare Architekturen, reaktive Entwicklung und GraphQL setzen,
```

Diese sollten effizient mit einer angepassten Auflösung und Qualität heruntergeladen werden.
Hier bieten sich Content Delivery Networks an, die diese Aufgabe gut erledigen.

# Skalierbarkeit

Skalierbarkeit wird häufig unter dem Aspekt betrachtet, dass hohe Zugriffszahlen auftreten, die sonst unüblich sind.
In der Regel gibt man sich aber ansonsten damit zufrieden, dass die Anwendung angemessen schnell reagiert.
Im Sinne von Green Software-Engineering muss Skalierbarkeit aber auch in die andere Richtung betrachtet werden.
Man sollte sich (auch aus Kostengründen) die Frage stellen, ob zwei Instanzen eines Microservice wirklich nötig sind, wenn gerade niemand auf diesen zugreift?
Reicht dann eine Instanz oder gar keine?

Das Thema Skalierbarkeit wurde in vielerlei Hinsicht durch das Aufkommen von Kubernetes (für die breite Masse) gelöst.
Dabei kann man ganz trivial eine Mindest- und eine Höchstzahl an Pods (Instanzen) pro Service definieren, die dann automatisch skalieren.
Entscheidend für die Technologie ist allerdings die Startzeit eines Microservice und da sind die auf "klassischen" Webframeworks basierenden Anwendungen leider recht langsam.

> Bei diesen Technologien ist mit einer Startzeit von über zehn Sekunden zu rechnen, in vielen Fällen sogar deutlich mehr (die Nutzung von GraalVM mit Spring Boot ermöglicht ebenfalls kurze Startzeiten, siehe [Running Spring Boot apps as GraalVM Native Images](https://blog.codecentric.de/en/2020/05/spring-boot-graalvm/))). 
> Das ist nicht wirklich kompatibel mit dem Skalierungsgedanken. 
> Natürlich spricht immer noch viel für den Einsatz von Spring Boot u.ä., bei diesem Aspekt sind aber andere Frameworks wie Quarkus, Micronaut und Ktor überlegen (siehe [Review of Microservices Frameworks](https://dzone.com/articles/not-only-spring-boot-a-review-of-alternatives)).

Reviewer: Ich hatte hier eher an so etwas gedacht:
```suggestion
Bei diesen Technologien ist mit einer Startzeit von über zehn Sekunden zu rechnen, in vielen Fällen sogar deutlich mehr. 
Das ist nicht wirklich kompatibel mit dem Skalierungsgedanken. 
Natürlich spricht immer noch viel für den Einsatz von Spring Boot u.ä., bei diesem Aspekt sind aber andere Frameworks wie Quarkus, Micronaut und Ktor überlegen (siehe [Review of Microservices Frameworks](https://dzone.com/articles/not-only-spring-boot-a-review-of-alternatives)).
Durch Ansätze wie die Nutzung von GraalVM kann dieses Problem in Spring Boot umgangen werden und es können Startzeiten im Millisekundenbereich erreicht werden (siehe [Running Spring Boot apps as GraalVM Native Images](https://blog.codecentric.de/en/2020/05/spring-boot-graalvm/)).
```
